### PR TITLE
Enabled LLM to return ToolRequestMessage even if tools is an empty list

### PR DESCRIPTION
**Pull request by maykcaldas**
Minor fixes:
-  refactored how `sum_logprobs` checks if `logprobs` are available in the `choices` to satisfy type-checking
- Updated tests to check if logprobs were correctly computed

### CODE REVIEW

File: tests/test_llms.py [hunk 1]
```diff
@@ -397,7 +401,11 @@ async def test_model(self, model_name: str) -> None:
             assert result.prompt_count > 0
             assert result.completion_count > 0
             assert result.cost > 0
-            assert result.logprob is None or result.logprob <= 0
+        if model.config["model_list"][0]["litellm_params"].get("logprobs"):
```
Comment (jamesbraza) [inline tests/test_llms.py:404]: Can you add a test for empty tools too?

File: llmclient/llms.py [hunk 2]
```diff
@@ -247,6 +245,7 @@ async def call(  # noqa: C901, PLR0915
             raise ValueError("Number of completions (n) must be >= 1.")
 
         # deal with tools
+        chat_kwargs["tools"] = tools  # Allows for empty tools list
```
Comment (jamesbraza) [inline llmclient/llms.py:248]: Can we do this in an `else` clause of the below line

File: tests/test_llms.py (hunk 1 above)
Comment (maykcaldas) [reply tests/test_llms.py:404]: Added in `TestTooling`

File: llmclient/llms.py (hunk 2 above)
Comment (maykcaldas) [reply llmclient/llms.py]: Moved it down

File: llmclient/llms.py [hunk 3]
```diff
@@ -562,6 +562,10 @@ async def check_rate_limit(self, token_count: float, **kwargs) -> None:
     @rate_limited
     async def acompletion(self, messages: list[Message], **kwargs) -> list[LLMResult]:
         # cast is necessary for LiteLLM typing bug: https://github.com/BerriAI/litellm/issues/7641
+        tools = None
+        if kwargs.get('tools', None) is not None:
+            tools = kwargs.pop('tools')
+
```
Comment (maykcaldas) [inline llmclient/llms.py:568]: The provider doesn't allow empty `tools`. So if it is an empty list, I am saving it but removing it from the kwargs so that it doesn't go to `acompletion`. Is there a better way of handling that?

File: llmclient/llms.py [hunk 4]
```diff
@@ -562,6 +562,10 @@ async def check_rate_limit(self, token_count: float, **kwargs) -> None:
     @rate_limited
     async def acompletion(self, messages: list[Message], **kwargs) -> list[LLMResult]:
         # cast is necessary for LiteLLM typing bug: https://github.com/BerriAI/litellm/issues/7641
+        tools = kwargs.get("tools")
+        if not tools:
+            kwargs.pop("tools", None)
+
```
Comment (maykcaldas) [inline llmclient/llms.py:568]: The llm provider doesn't allow requests with empty `tools` list. So I am removing it from kwargs if that's the case.

File: tests/test_llms.py [hunk 5]
```diff
@@ -628,6 +636,32 @@ def double(x: int) -> int:
...
 
+    @pytest.mark.asyncio
+    @pytest.mark.parametrize(
+        "tools",
+        [
+            pytest.param([], id="empty-tools"),
+            pytest.param(None, id="no-tools"),
+        ],
+    )
+    async def test_empty_tools(self, tools) -> None:
```
Comment (jamesbraza) [inline tests/test_llms.py:647]: Can we add a VCR to this

File: llmclient/llms.py (hunk 4 above)
Comment (jamesbraza) [reply llmclient/llms.py:568]: Which LLM provider? Can you document this with a code comment?

Comment (maykcaldas) [reply llmclient/llms.py:569]: Specifically LiteLLM. But I think that's because they follow OpenAI's interface (which also don't allow empty tools calls). I added a test with anthropic models to the test. But all of them pass through LiteLLM anyway

File: tests/test_llms.py [hunk 6]
```diff
@@ -628,6 +636,37 @@ def double(x: int) -> int:
...
+            pytest.param(
+                [], CommonLLMNames.ANTHROPIC_TEST.value, id="Anthropic-empty-tools"
+            ),
+            pytest.param(
+                None, CommonLLMNames.ANTHROPIC_TEST.value, id="Anthropic-no-tools"
+            ),
+        ],
+    )
+    @pytest.mark.vcr
+    async def test_empty_tools(self, tools, model_name) -> None:
```
[the remark this answers is not in the document]
Comment (maykcaldas) [reply tests/test_llms.py]: Here we always use an empty list. That's the point of this test. So, I typed it with`list` only. Without specifying the list content type

File: llmclient/llms.py [hunk 7]
```diff
@@ -561,6 +561,12 @@ async def check_rate_limit(self, token_count: float, **kwargs) -> None:
 
     @rate_limited
     async def acompletion(self, messages: list[Message], **kwargs) -> list[LLMResult]:
+        tools = kwargs.get("tools")
+        if not tools:
+            # OpenAI, Anthropic and pottentially other LLM providers
+            # don't allow empty tool_calls lists, so downcast empty
```
Comment (jamesbraza) [inline llmclient/llms.py:567]: Note it's not a downcast, we are actually removing it. Can you adjust this comment

**Review by jamesbraza — APPROVED** [review]
Nice work, thanks for this
